[core] Fix XMLRenderer encoding issues

adangel opened the pull request:
## Describe the PR

This PR tries to fix #2615 . In order to use the correct encoding through the rendering process, the new API `Renderer::setReportFile` is added.

I'm not entirely sure about CPD renderer, whether we have a similar issue there. But as far as I can see, only the default platform encoding is used, there is no way to change the encoding. There is also no way to specify a output file as a command line option, so CPD renderers always render to Sysout and hence should use the default platform encoding.

## Related issues

- Fixes #2615 

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [x] Added (in-code) documentation (if needed).

adangel commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/XMLRenderer.java:
```diff
@@ -119,7 +121,8 @@ private Element addFilesToDuplicationElement(Document doc, Element duplication,
             final Element file = doc.createElement("file");
             file.setAttribute("line", String.valueOf(mark.getBeginLine()));
-            file.setAttribute("path", mark.getFilename());
+            String filenameXml10 = StringEscapeUtils.unescapeXml(StringEscapeUtils.escapeXml10(mark.getFilename()));
```
StringEscapeUtils escape basic chars like "<" and remove invalid chars like U+000C. But the DOM impl already escapes "<" by itself in attribute values, but doesn't remove invalid chars. In order to avoid double escaping, this escapes it once with StringEscapeUtils and unescapes it again - result is, that only invalid chars are removed (no escaping).

adangel commented on pmd-core/src/main/java/net/sourceforge/pmd/renderers/XMLRenderer.java:
```diff
@@ -179,6 +182,41 @@ private void createTimestampAttr(StringBuilder buffer) {
+            StringBuilder sb = new StringBuilder(result);
+            for (int i = 0; i < sb.length(); i++) {
+                char c = sb.charAt(i);
+                // surrogate characters are not allowed in XML
```
this and a non unicode based encoding like ISO-8859-1 or cp1252 cannot encode codepoints > 255.
Hm... so, we should probably escape chars additionally as soon as `c > 0xff`, because then you need unicode...

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This changeset introduces 1062...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 106260 new violations, 0 new errors and 0 new configuration errors,<br />
removes 122673 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/e288a6493600436299dd07562290d81b/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

adangel commented:
This PR showed actually a bug in pmd-regression-tester: With this PR, we are now escaping certain characters in XML, which we previously didn't:

before:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<pmd xmlns="http://pmd.sourceforge.net/report/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/report/2.0.0 http://pmd.sourceforge.net/report_2_0_0.xsd"
    version="6.24.0" timestamp="2020-07-05T21:26:00.982">
<file name="/home/andreas/Downloads/chunk-diff-issue-2615/src/TestFile.java">
<violation beginline="2" endline="2" begincolumn="21" endcolumn="35" rule="MethodArgumentCouldBeFinal" ruleset="Code Style" class="TestFile" method="bar" variable="fileName" externalInfoUrl="https://pmd.github.io/pmd-6.24.0/pmd_rules_java_codestyle.html#methodargumentcouldbefinal" priority="3">
Parameter 'fileName' is not assigned and could be declared final
</violation>
</file>
</pmd>
```

after:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<pmd xmlns="http://pmd.sourceforge.net/report/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/report/2.0.0 http://pmd.sourceforge.net/report_2_0_0.xsd"
    version="6.26.0-SNAPSHOT" timestamp="2020-07-05T21:26:32.079">
<file name="/home/andreas/Downloads/chunk-diff-issue-2615/src/TestFile.java">
<violation beginline="2" endline="2" begincolumn="21" endcolumn="35" rule="MethodArgumentCouldBeFinal" ruleset="Code Style" class="TestFile" method="bar" variable="fileName" externalInfoUrl="https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#methodargumentcouldbefinal" priority="3">
Parameter &apos;fileName&apos; is not assigned and could be declared final
</violation>
</file>
</pmd>
```

pmd-regression-tester doesn't parse the character data correctly - with entities, the character data between tags (where the description is), is now split into multiple chunks of character data, which needs to be accumulated. The parser currently only takes the last character data. Hence the difference in the violations, e.g. before: `Parameter 'fileName' is not assigned and could be declared final`, after: `is not assigned and could be declared final`.

I'll fix that tomorrow.

adangel commented:
I think, this is ready now.
I refactored the XMLRenderer for PMD to use XMLStreamWriter instead of manually creating XML via StringBuilder... 
I learned that the standard XML libraries mostly work correctly, at least in terms of encoding - the only thing to keep in mind is: it works only, if you provide a OutputStream - if you provide a Writer, then now charset specific encoding is happening (like dealing with unmappable characters). Invalid characters are still kept unfortunately, hence I created a new StringUtil method....

I've added the experimental API `Renderer::setReportFile` which is good for now. But for PMD 7, we should rethink the API: Maybe we can say, that each renderer works with a OutputStream and has to know the encoding (falling back to platform default). It would be nice, to have the Renderer API decoupled from File - at least in unit tests, you often just want to render into memory.

adangel commented:
refs #1962

oowekyala commented on pmd-core/src/main/java/net/sourceforge/pmd/renderers/Renderer.java:
```diff
@@ -163,4 +164,17 @@
+     * @param reportFilename the filename (optional).
+     */
+    @Experimental
+    void setReportFile(String reportFilename);
```
In a mature API, I think this should take a `Path` argument instead of a string. But then, everything should be updated to be consistent
